File: python/removeLoopFromLinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def creatte_linkedList(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        last = self.head
        while last.next:
            last = last.next
        
        last.next = new_node

    def remove_loop(self):
        slow = self.head
        fast = self.head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

            if slow == fast:
                break
        
        if slow != fast:
            return

        start = self.head
        if slow == start:
            while fast.next != self.head:
                fast = fast.next 
        
        else:
            while start.next != slow.next:
                start = start.next
                slow = slow.next

        slow.next = None
        print("Loop removed from the linked list")
    
    def print_list(self):
        current = self.head
        while current:
            print(current.data, end=" ")
            current = current.next
        print()

def main():
    llist = LinkedList()
    llist.creatte_linkedList(1)
    llist.creatte_linkedList(2)
    llist.creatte_linkedList(3)
    llist.creatte_linkedList(4)
    llist.creatte_linkedList(5)

    # Creating a loop for testing
    llist.head.next.next.next.next.next = llist.head.next

    print("Linked List before removing loop:")
    llist.print_list()

    llist.remove_loop()

    print("Linked List after removing loop:")
    llist.print_list()

if __name__ == "__main__":
    main()
    

File: javaFile/remove_loop_from_linkedList.java
class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    Node head;

    void append(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            return;
        }
        Node current = head;
        while (current.next != null) {
            current = current.next;
        }
        current.next = newNode;
    }

    void createLoop(int position) {
        Node temp = head;
        while (temp.next != null) {
            temp = temp.next;
        }
        temp.next = head.next.next;
    }

    void removeLoop() {
        Node temp1 = head;
        Node temp2 = head;

        // Detect loop using Floyd's cycle detection algorithm
        while ((temp2 != null) || (temp2.next != null)) {
            temp1 = temp1.next;
            temp2 = temp2.next.next;

            // If loop exists, break
            if (temp1 == temp2) {
                break;
            }
        }
        // If no loop exists, return
        if (temp1 == temp2) {
            temp1 = head;
            while (temp1 != temp2) {
                temp1 = temp1.next;
                temp2 = temp2.next;
            }
            // last node
            while (temp1.next != temp2) {
                temp1 = temp1.next;

            }
            temp2.next = null; // remove loop

        }
    }

    void printList() {
        Node current = head;
        while (current != null) {
            System.out.print(current.data + " ");
            current = current.next;
        }
        System.out.println();
    }
}

public class remove_loop_from_linkedList {

    public static void main(String[] args) {
        LinkedList list = new LinkedList();
        list.append(1);
        list.append(2);
        list.append(3);
        list.append(4);
        list.append(5);

        list.createLoop(2); // Create a loop for testing
        list.removeLoop(); // Remove the loop
        list.printList(); // Print the list after removing the loop

        
    }
}
